WalletWasabi.Tests/RegressionTests/CancelTests.cs

```diff
@@ -0,0 +1,259 @@
+				}
+			}
+
+			#region HasChange
+
+			Money amountToSend = wallet.Coins.Where(x => x.IsAvailable()).Sum(x => x.Amount) / 2;
+			var externalAddr = await rpc.GetNewAddressAsync(CancellationToken.None);
+			var txToCancel = wallet.BuildTransaction(password, new PaymentIntent(externalAddr, amountToSend, label: "bar"), FeeStrategy.SevenDaysConfirmationTargetStrategy, allowUnconfirmed: true);
+			await broadcaster.SendTransactionAsync(txToCancel.Transaction);
+
+			var cancellingTx = wallet.CancelTransaction(txToCancel.Transaction);
+
+			var changeOut = Assert.Single(txToCancel.InnerWalletOutputs);
+			var cancelOut = Assert.Single(cancellingTx.InnerWalletOutputs);
+			Assert.Equal(changeOut.ScriptPubKey, cancelOut.ScriptPubKey);
+
+			Assert.Single(txToCancel.OuterWalletOutputs);
+			Assert.Empty(cancellingTx.OuterWalletOutputs);
+
+			Assert.True(txToCancel.Fee < cancellingTx.Fee);
+			Assert.NotEqual(0, txToCancel.FeePercentOfSent);
+			Assert.Equal(0, cancellingTx.FeePercentOfSent);
+
+			Assert.Equal(txToCancel.SpentCoins, cancellingTx.SpentCoins);
+
+			await broadcaster.SendTransactionAsync(cancellingTx.Transaction);
+			Assert.False(txToCancel.Transaction.IsReplacement);
+			Assert.False(txToCancel.Transaction.IsCpfp);
+			Assert.False(txToCancel.Transaction.IsCancellation);
+			Assert.True(cancellingTx.Transaction.IsReplacement);
+			Assert.False(cancellingTx.Transaction.IsCpfp);
+			Assert.True(cancellingTx.Transaction.IsCancellation);
+
+			await rpc.GenerateAsync(1);
+
+			#endregion HasChange
+
+			#region NoChange
+
+			externalAddr = await rpc.GetNewAddressAsync(CancellationToken.None);
+			txToCancel = wallet.BuildChangelessTransaction(externalAddr, "foo", new FeeRate(1m), cancellingTx.InnerWalletOutputs);
+			await broadcaster.SendTransactionAsync(txToCancel.Transaction);
+
+			cancellingTx = wallet.CancelTransaction(txToCancel.Transaction);
+
+			Assert.Empty(txToCancel.InnerWalletOutputs);
+			Assert.Single(cancellingTx.InnerWalletOutputs);
+
+			Assert.Single(txToCancel.OuterWalletOutputs);
+			Assert.Empty(cancellingTx.OuterWalletOutputs);
+
+			Assert.True(txToCancel.Fee < cancellingTx.Fee);
+			Assert.NotEqual(0, txToCancel.FeePercentOfSent);
+			Assert.Equal(0, cancellingTx.FeePercentOfSent);
+
+			Assert.Equal(txToCancel.SpentCoins, cancellingTx.SpentCoins);
+
+			await broadcaster.SendTransactionAsync(cancellingTx.Transaction);
+			Assert.False(txToCancel.Transaction.IsReplacement);
+			Assert.False(txToCancel.Transaction.IsCpfp);
+			Assert.False(txToCancel.Transaction.IsCancellation);
+			Assert.True(cancellingTx.Transaction.IsReplacement);
+			Assert.False(cancellingTx.Transaction.IsCpfp);
+			Assert.True(cancellingTx.Transaction.IsCancellation);
+
+			#endregion NoChange
+
+			#region CantCancel
+
+			// Can't cancel cancelled transacion.
+			Assert.Throws<InvalidOperationException>(() => wallet.CancelTransaction(cancellingTx.Transaction));
```

Comment:
> This doesn't fail anymore

It does now, so I won't investigate.

> your comment is incorrect

Fixed.

> On another note, maybe we should also make sure that we can't cancel 2 times the same transaction? The PR doesn't handle that case.

Interesting consideration. However, I believe that implementing this change would not be beneficial for the code.
**Costs.** Handling it is complex: It would require the addition of a new state to the smart transaction, setting it in the transaction processor, and managing the storage of canceled transactions, which could lead to further complications.

**Benefits.** Moreover, it is important to note that such a change would not have any impact on the production environment. These transactions are not visible in the UI history, making it impossible to access them. Even if we were to expand the scope of the PR and integrate with RPC, the second cancellation of the transaction's broadcast would still fail.